Plugin Extensions: DataSource Configuration Components

shelldandy opened the pull request:
This PR adds plugin extension points to DataSource configuration components, enabling plugins to register custom actions and status links in the datasource configuration interface. The implementation focuses on providing entry points for drilldown apps and healthcheck interfaces.

<img width="795" height="598" alt="image" src="https://github.com/user-attachments/assets/6c541b94-d733-4890-bc32-c45f203a4acc" />

<img width="921" height="266" alt="image" src="https://github.com/user-attachments/assets/b91a041e-6bb1-44e7-816e-b17045b04b94" />

Introduces two new extension points: `DataSourceConfigActions` and `DataSourceConfigStatus`
Updates `EditDataSourceActions` and `DataSourceTestingStatus` components to consume plugin extensions

## Overview
This specification defines the addition of plugin extension points to DataSource configuration components, allowing plugins to register custom actions and status links in the datasource configuration interface.

tl;dr We want to add more entry points for the drilldown apps, healthcheck to drilldown interface is added

Only limited to the following plugins:

```
'grafana-lokiexplore-app',
  'grafana-exploretraces-app',
  'grafana-metricsdrilldown-app',
  'grafana-pyroscope-app',
  'grafana-monitoring-app',
  'grafana-troubleshooting-app',
```

Supports: https://github.com/grafana/metrics-drilldown/issues/424?reload=1?reload=1

shelldandy committed:
Merge branch 'main' of github.com:grafana/grafana into feat/healthcheck-2-gmd

shelldandy committed:
add the plan

shelldandy committed:
feat: Add plugin extension points for DataSource configuration components

- Add DataSourceConfigActions and DataSourceConfigStatus extension points
- Add PluginExtensionDataSourceConfigActionsContext and PluginExtensionDataSourceConfigStatusContext types
- Update EditDataSourceActions.tsx with plugin extension support and allowlist filtering
- Update DataSourceTestingStatus.tsx with dual extension point support (new + backward compatible)
- Create getDataSourceExtensionConfigs.tsx with example core extensions
- Update getCoreExtensionConfigurations.ts to include datasource extensions
- Implement grafana-owned plugin allowlist for security and quality control
- Support context-aware extensions based on datasource type and testing status
- Maintain backward compatibility with existing DataSourceConfigErrorStatus extension point
- Add proper TypeScript types and export them in index.ts

This enables grafana-owned plugins to provide contextual actions and status-specific integrations
within the datasource configuration workflow while maintaining security through plugin filtering.

shelldandy committed:
docs: Update specs.md with implementation status and lessons learned

- Add comprehensive implementation status section with completed features
- Document critical lessons learned during implementation:
  * Translation limitations in extension configurations
  * Plugin security through allowlist filtering
  * Extension registration timing constraints
  * TypeScript context type patterns
- Update implementation checklist with actual completion status
- Add detailed next steps and recommendations
- Include plugin development guidelines (do's and don'ts)
- Document architectural insights and patterns discovered
- Transform specs from theoretical design to practical implementation guide

This update serves as both historical record and developer guide for future
plugin extension work in Grafana.

shelldandy committed:
Add comprehensive tests for DataSource plugin extensions

- Add EditDataSourceActions.test.tsx with 12 tests covering:
  * Core Grafana actions functionality and permissions
  * Plugin extension rendering and interaction
  * Plugin allowlist filtering and security
  * Context passing and URL generation

- Add getDataSourceExtensionConfigs.test.tsx with 14 tests covering:
  * Core extension configurations for both extension points
  * Context filtering logic (datasource type, severity)
  * Error handling and graceful failures
  * Extension structure validation and type safety

Total: 26 tests passing, providing comprehensive coverage of the
DataSourceConfigActions and DataSourceConfigStatus extension points
implemented in previous commits.

shelldandy committed:
Update specs.md to reflect completed test implementation

- Mark unit testing phase as complete (26 tests added)
- Add comprehensive Test Implementation section documenting:
  * Test file details and coverage areas
  * Testing challenges and solutions discovered
  * Performance metrics and execution strategy
  * Command examples for running tests
- Update Success Metrics with concrete testing validation
- Update Next Steps to reflect testing completion
- Document new test files in Files Modified section

The plugin extensions implementation now has complete unit test
coverage for both DataSourceConfigActions and DataSourceConfigStatus
extension points with 26 passing tests.

shelldandy committed:
Merge branch 'main' of github.com:grafana/grafana into feat/healthcheck-2-gmd

shelldandy committed:
rm specs file

shelldandy committed:
Merge branch 'main' of github.com:grafana/grafana into feat/healthcheck-2-gmd

shelldandy committed:
lint

shelldandy committed:
Merge branch 'main' of github.com:grafana/grafana into feat/healthcheck-2-gmd

shelldandy committed:
refactor: replace 'any' types with proper TypeScript types in datasource extension tests

shelldandy committed:
tsc

shelldandy committed:
prettier: fix

shelldandy committed:
Fix datasource extension path format to resolve CI test failure

The troubleshooting link extension was using an invalid path format
'/docs/troubleshooting/datasources' which caused console.error during
test execution. Updated to proper Grafana core extension path format
'/a/grafana/docs/troubleshooting/datasources' to comply with plugin
extension validation rules.

shelldandy committed:
fix: DataSourceTestingStatus test failures

- Fix incorrect data-testid expectations to use proper e2e selectors
- Add contextSrv mock for hasAccessToExplore() dependency
- Create proper plugin links mock handling different extension points
- Update plugin link tests to use allowed plugin IDs
- Fix test logic to match actual component behavior for status/error links
- Add test coverage for non-allowed plugin filtering

All 10 tests now pass successfully.

shelldandy committed:
lint fix

shelldandy committed:
refactor: replace any types with proper TypeScript types in DataSourceTestingStatus test

- Replace any[] with PluginExtensionLink[] for statusLinks and errorLinks parameters
- Replace any with UsePluginLinksOptions for setPluginLinksHook callback parameter
- Add proper imports for PluginExtensionLink and UsePluginLinksOptions
- Improve type safety and IntelliSense support in tests

leventebalogh commented on public/app/features/datasources/extensions/getDataSourceExtensionConfigs.test.tsx:
```diff
@@ -0,0 +1,274 @@
+      expect(troubleshootingConfig?.configure).toBeDefined();
+    });
+  });
+});
```
Hmm, what's the goal of these tests? The `getDataSourceExtensionConfigs()` function is mostly just returning a configuration data object (with some minimal business logic). By defining the config values twice (once in the tests and once in the config object), the tests will break every time the config values change, making them more like an overhead in my opinion.

shelldandy committed:
Merge branch 'main' of github.com:grafana/grafana into feat/healthcheck-2-gmd

shelldandy committed:
rm brittle tests

shelldandy committed:
DataSource: Replace individual buttons with dropdown menu in EditDataSourceActions

- Convert LinkButton components to dropdown with Menu.Item components
- Add conditional rendering: show simple button when only one action, dropdown when multiple
- Use LinkButton for navigation and Button with dropdown icon for menu trigger
- Replace Trans components with t() function calls for Menu.Item labels
- Prevent PageHeader crowding when multiple plugin extensions are registered
- Maintain all existing functionality and tracking calls

shelldandy committed:
refactor: use configure() for dynamic datasource extension URLs

- Replace onClick handler with configure() function for dynamic path generation
- Use context to build monitoring tool URL with datasource UID
- Follows extension system design patterns for declarative configuration
- Removes manual window.open() in favor of automatic path handling

shelldandy committed:
refactor: use DataSourceConfigErrorStatus extension point for troubleshooting guide

Use the specific DataSourceConfigErrorStatus extension point instead of
DataSourceConfigStatus with conditional logic in configure function.
This simplifies the code by leveraging the right abstraction level.

shelldandy committed:
extract i18n

shelldandy committed:
lint:fix

shelldandy committed:
Revert "lint:fix"

This reverts commit 23cdb39672a289e251910983747e412f641bdc16.

shelldandy committed:
lint:fix

shelldandy committed:
fix: mock plugin extensions registry in variables utils test

Fixes console.error calls during test execution caused by plugin extension
registration failures. The test was failing on CI because the extensions
registry was trying to register link extensions with invalid configurations
during module import.

Solution: Mock the entire plugin extensions registry setup to prevent
side effects from interfering with the test execution.

shelldandy committed:
fix: mock plugin extensions registry in explore query test

Fixes console.error calls during test execution caused by plugin extension
registration failures. The test was failing on CI because the extensions
registry was trying to register link extensions with invalid configurations
during module import.

This applies the same fix as the variables utils test - mocking the entire
plugin extensions registry setup to prevent side effects from interfering
with the test execution.

shelldandy committed:
better fix

shelldandy committed:
Fix EditDataSourceActions test by mocking utils and handling dropdown UI

- Add missing mock for constructDataSourceExploreUrl from ../utils
- Update tests to interact with dropdown menu structure instead of expecting direct text
- Fix test logic to match component behavior: Actions dropdown when user has explore rights

shelldandy committed:
lint:fix

shelldandy commented on public/app/features/datasources/extensions/getDataSourceExtensionConfigs.tsx:
```diff
@@ -0,0 +1,56 @@
+
+          // Return dynamic path with context
+          return {
+            path: `https://monitoring-tool.com/datasource/${context.dataSource.uid}`,
```
Oh, it's just the example but it can be a drill down app url or something

leventebalogh replied:
Yeah, I don't think we should leave any dead code as examples in there, it's quite confusing. (I mean dead code, because here we register these as core grafana extensions for these extension points, but we filter them out later on at the extension points - with the plugin id filters).

shelldandy replied:
Ah i see what you mean, would it work better if i add a metrics-drilldown url for the example?

jackw commented on public/app/features/datasources/extensions/getDataSourceExtensionConfigs.tsx:
```diff
@@ -0,0 +1,56 @@
+        icon: 'question-circle',
+        // eslint-disable-next-line @grafana/i18n/no-untranslated-strings
+        category: 'Help',
+        path: '/a/grafana/docs/troubleshooting/datasources',
```
Where does this path resolve to? It doesn't look like a link to a page in grafana core nor a plugin.

jackw commented on public/app/features/plugins/extensions/getCoreExtensionConfigurations.ts:
```diff
@@ -1,6 +1,7 @@
 
 export function getCoreExtensionConfigurations(): PluginExtensionAddedLinkConfig[] {
-  return [...getExploreExtensionConfigs()];
+  return [...getExploreExtensionConfigs(), ...getDataSourceExtensionConfigs()];
```
Can you explain why we need to add this to coreExtensionConfigs please?

jackw commented on public/app/features/datasources/components/EditDataSourceActions.tsx:
```diff
@@ -11,17 +12,66 @@ interface Props {
   uid: string;
 }
 
+const allowedPluginIds = [
```
This list appears to be duplicated in another component? Can we use a single source of truth?

shelldandy replied:
Roger that, moving them to constants

leventebalogh commented on public/app/features/plugins/extensions/getCoreExtensionConfigurations.ts:
```diff
@@ -1,6 +1,7 @@
 
 export function getCoreExtensionConfigurations(): PluginExtensionAddedLinkConfig[] {
-  return [...getExploreExtensionConfigs()];
+  return [...getExploreExtensionConfigs(), ...getDataSourceExtensionConfigs()];
```
I think we can remove this (we don't have any core extensions yet if I understand correctly).

shelldandy replied:
Roger that. I misunderstood how that array worked. It can all be removed

leventebalogh commented on public/app/features/datasources/components/EditDataSourceActions.tsx:
```diff
@@ -51,6 +77,21 @@ export function EditDataSourceActions({ uid }: Props) {
+          >
+            {link.title}
+          </LinkButton>
+        ))}
```
I think this is still relevant?

shelldandy replied:
the actions are now inside a dropdown, the code looked similar but that was because i wrapped them inside a `actionsMenu` component to then render them later in the return function

<img width="1106" height="890" alt="image" src="https://github.com/user-attachments/assets/8380c784-5022-4c85-b3d7-3d02e5fd26e1" />

shelldandy committed:
refactor: extract duplicated allowedPluginIds to single source of truth

- Add ALLOWED_DATASOURCE_EXTENSION_PLUGINS constant to constants.ts
- Remove duplicate allowedPluginIds arrays from DataSourceTestingStatus and EditDataSourceActions
- Update both components to import and use the shared constant
- Improves maintainability and ensures consistency across datasource config components

shelldandy committed:
refactor: update datasource extension link configurations

- Change title and description for the metrics drilldown link to improve clarity
- Update the path for the metrics drilldown link to reflect the new routing structure
- Remove the troubleshooting guide link configuration to streamline the extension options

shelldandy committed:
removing unused config

shelldandy committed:
better dropdown

shelldandy committed:
fix: update EditDataSourceActions tests to match component implementation

- Fixed tests to expect core actions as separate buttons instead of dropdown
- Updated plugin extension tests to use 'Extensions' dropdown instead of 'Actions'
- Added test for Extensions dropdown visibility logic
- All 12 tests now passing

shelldandy committed:
lint:fix

shelldandy committed:
Merge branch 'main' of github.com:grafana/grafana into feat/healthcheck-2-gmd

leventebalogh commented on public/app/features/datasources/components/EditDataSourceActions.tsx:
```diff
@@ -15,6 +17,34 @@ export function EditDataSourceActions({ uid }: Props) {
+
+  const actionsMenu = (
+    <Menu>
+      {hasActions &&
```
Do we need the `hasActions` check here? (We are also checking it at line 84)

leventebalogh commented on public/app/features/datasources/constants.ts:
```diff
@@ -10,3 +10,15 @@ export const DATASOURCES_ROUTES: DataSourcesRoutes = {
 } as const;
+
+/**
+ * Plugin IDs that are allowed to contribute extensions to datasource configuration pages
```
I think this could be a bit misleading, as the `EditDatasource` page allows only `grafana-pdc-app` and `grafana-auth-app`: https://github.com/grafana/grafana/blob/main/public/app/features/datasources/components/EditDataSource.tsx#L271

shelldandy committed:
remove extra check

shelldandy committed:
Merge branch 'main' of github.com:grafana/grafana into feat/healthcheck-2-gmd

shelldandy committed:
Update ALLOWED_DATASOURCE_EXTENSION_PLUGINS documentation to clarify plugin contributions to datasource configuration pages

shelldandy committed:
feat: Convert Explore data button to dropdown when extensions are available

- When hasActions is empty: shows normal Explore data and Build dashboard buttons
- When hasActions is not empty: converts Explore data to dropdown with angle-down icon
- First dropdown item: 'Open in Explore View'
- Following items: extension links
- Build dashboard button remains unchanged
- Maintains existing tracking functionality and internationalization

shelldandy committed:
test: Update EditDataSourceActions tests for new dropdown implementation

- Update tests to work with new Explore data dropdown that includes extensions
- Replace references to 'Extensions' button with 'Explore data' dropdown
- Add test for 'Open in Explore View' menu item in dropdown
- Verify correct behavior when extensions are present vs absent
- All 13 tests now passing

shelldandy committed:
test: Add comprehensive test coverage for dropdown explore functionality

- Add test to verify explore action href when extensions are present
- Ensure both dropdown and direct link behaviors are tested
- Complete test coverage for new dropdown implementation

shelldandy committed:
translation

shelldandy committed:
compass icon for the explore item

shelldandy committed:
feat: refactor AlertSuccessMessage to accept extension links as props

- Move extension links from inline definition to props for better reusability
- Add proper TypeScript support for PluginExtensionLink types
- Improve component architecture by separating concerns
- Add extensionLinks styling to getStyles function for consistency
- Support both real extension links and custom link rendering in success messages

This improves the component's flexibility while maintaining type safety.

shelldandy committed:
Merge branch 'main' of github.com:grafana/grafana into feat/healthcheck-2-gmd

shelldandy committed:
lint:fix

shelldandy committed:
Merge branch 'main' of github.com:grafana/grafana into feat/healthcheck-2-gmd

jackw reviewed (approved):
LGTM! 🚀

shelldandy committed:
Merge branch 'main' of github.com:grafana/grafana into feat/healthcheck-2-gmd
